File: lsd-interceptors/src/main/java/io/lsdconsulting/interceptors/http/LsdFeignLoggerInterceptor.java
package io.lsdconsulting.interceptors.http;

import com.lsd.core.LsdContext;
import feign.Logger;
import feign.Request;
import feign.Response;
import feign.Util;
import io.lsdconsulting.interceptors.common.Headers;
import io.lsdconsulting.interceptors.http.common.HttpInteractionHandler;
import org.springframework.http.HttpStatus;

import java.io.IOException;
import java.util.List;
import java.util.Optional;

/**
 * Intercepts Feign {@link Request} and {@link Response} messages to add them to the {@link LsdContext} class.
 * <p>
 * (This allows lsd-core to display them on the sequence diagrams).
 */
public class LsdFeignLoggerInterceptor extends Logger.JavaLogger {
    public static final String EXTRACT_PATH = "https?://.*?(/.*)";

    protected final List<HttpInteractionHandler> handlers;

    public LsdFeignLoggerInterceptor(List<HttpInteractionHandler> handlers) {
        super(LsdFeignLoggerInterceptor.class);
        this.handlers = handlers;
    }

    @Override
    protected void logRequest(String configKey, Level level, Request request) {
        super.logRequest(configKey, level, request);
        captureRequestInteraction(request);
    }

    @Override
    protected Response logAndRebufferResponse(String configKey, Level logLevel, Response response, long elapsedTime) throws IOException {
        super.logAndRebufferResponse(configKey, logLevel, response, elapsedTime);
        String body = extractResponseBodyToString(response);
        captureResponseInteraction(response, body);
        return resetBodyData(response, body.getBytes());
    }

    private void captureRequestInteraction(Request request) {
        Optional<byte[]> bodyData = Optional.ofNullable(request.body());
        String body = bodyData.map(String::new).orElse("");
        String path = derivePath(request.url());
        var headers = Headers.singleValueMap(request.headers());

        handlers.forEach(handler ->
                handler.handleRequest(request.httpMethod().name(), headers, path, body));
    }

    private void captureResponseInteraction(Response response, String body) {
        String path = derivePath(response.request().url());
        var requestHeaders = Headers.singleValueMap(response.request().headers());
        var responseHeaders = Headers.singleValueMap(response.headers());

        handlers.forEach(handler ->
                handler.handleResponse(deriveStatus(response.status()), requestHeaders, responseHeaders, path, body));
    }

    private String deriveStatus(int code) {
        Optional<HttpStatus> httpStatus = Optional.ofNullable(HttpStatus.resolve(code));
        return httpStatus.map(HttpStatus::toString)
                .orElse(String.format("<unresolved status:%s>", code));
    }

    private String derivePath(String url) {
        return url.replaceAll(EXTRACT_PATH, "$1");
    }

    private String extractResponseBodyToString(Response response) throws IOException {
        byte[] bytes = Util.toByteArray(response.body().asInputStream());
        return new String(bytes);
    }

    private Response resetBodyData(Response response, byte[] bodyData) {
        return response.toBuilder().body(bodyData).build();
    }
}

File: lsd-interceptors/src/main/kotlin/io/lsdconsulting/interceptors/http/LsdFeignLoggerInterceptor.kt
package io.lsdconsulting.interceptors.http

import feign.Logger.JavaLogger
import feign.Request
import feign.Response
import feign.Util
import io.lsdconsulting.interceptors.common.singleValueMap
import io.lsdconsulting.interceptors.http.common.HttpInteractionHandler
import org.springframework.http.HttpStatus
import java.io.IOException
import java.time.Duration
import java.util.*
import java.util.function.Consumer

const val EXTRACT_PATH = "https?://.*?(/.*)"

/**
 * Intercepts Feign [Request] and [Response] messages to add them to the [com.lsd.core.LsdContext] class.
 *
 *
 * (This allows lsd-core to display them on the sequence diagrams).
 */
open class LsdFeignLoggerInterceptor(val handlers: List<HttpInteractionHandler>) : JavaLogger(
    LsdFeignLoggerInterceptor::class.java
) {
    override fun logRequest(configKey: String, level: Level, request: Request) {
        super.logRequest(configKey, level, request)
        captureRequestInteraction(request)
    }

    @Throws(IOException::class)
    override fun logAndRebufferResponse(
        configKey: String,
        logLevel: Level,
        response: Response,
        elapsedTime: Long
    ): Response {
        super.logAndRebufferResponse(configKey, logLevel, response, elapsedTime)
        val body = extractResponseBodyToString(response)
        captureResponseInteraction(response, body, Duration.ofMillis(elapsedTime))
        return resetBodyData(response, body.toByteArray())
    }

    private fun captureRequestInteraction(request: Request) {
        val bodyData = Optional.ofNullable(request.body())
        val body = bodyData.map { bytes: ByteArray? ->
            String(
                bytes!!
            )
        }.orElse("")
        val path = derivePath(request.url())
        val headers = singleValueMap(request.headers())
        handlers.forEach(Consumer { handler: HttpInteractionHandler ->
            handler.handleRequest(
                request.httpMethod().name,
                headers,
                path,
                body
            )
        })
    }

    private fun captureResponseInteraction(response: Response, body: String, duration: Duration) {
        val path = derivePath(response.request().url())
        val requestHeaders = singleValueMap(response.request().headers())
        val responseHeaders = singleValueMap(response.headers())
        handlers.forEach(Consumer { handler: HttpInteractionHandler ->
            handler.handleResponse(
                deriveStatus(response.status()),
                requestHeaders,
                responseHeaders,
                path,
                body,
                duration
            )
        })
    }

    private fun deriveStatus(code: Int): String =
        HttpStatus.resolve(code)?.let { obj: HttpStatus -> obj.toString() }
            ?: String.format("<unresolved status:%s>", code)

    private fun derivePath(url: String): String = url.replace(EXTRACT_PATH.toRegex(), "$1")

    @Throws(IOException::class)
    private fun extractResponseBodyToString(response: Response): String {
        return response.body()?.let { String(Util.toByteArray(it.asInputStream())) } ?: ""
    }

    private fun resetBodyData(response: Response, bodyData: ByteArray): Response {
        return response.toBuilder().body(bodyData).build()
    }
}
